### ISSUE TITLE
omit_local_variable_types false positive

### ISSUE THREAD

**Issue by davidmorgan** [dart-lang/linter#2547]
It looks like the lint considers `Foo<dynamic>` and `Foo<Object>` to be the same, so it doesn't allow a left hand side type to 'cast' between the two:

```
class A<T> {
  T _value;
  A(this._value);
 
  T get value => _value;
}

class Foo {
  String foo;
 
  Foo(this.foo);
}

A<Object> b() {
  return A<Foo>(Foo('hello world'));
}

void main() async {
  // This gives #omit_local_variable_types
  // final A a = b();
  // This gives The getter 'foo' isn't defined for the class 'Object'.
  final a = b();
  final s = a.value.foo;
  print(s);
}
```

**Comment by srawlins** [comment]
I believe this is a duplicate of https://github.com/dart-lang/linter/issues/3469, which has lots of discussion.

**Comment by davidmorgan** [comment]
@srawlins I don't think it's a duplicate:

- This issue is about a problem with the lint; it treats `dynamic` and `Object` as the same, so it gives a false positive
- #3469 is about a problem with the quick fix, it removes type information from the LHS without putting it back on the RHS
Thanks.

**Comment by srawlins** [comment]
Good catch.
